onboard/merge_iterator.go

```diff
@@ -0,0 +1,170 @@
+package onboard
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/treeverse/lakefs/catalog/rocks"
+)
+
+// prefixMergeIterator takes inventory iterator and commit iterator and merges them.
+// If an entry's prefix is in `prefixes`, it would be taken from invIt.
+// No entry from committedIt would be returned if its prefix is in prefixes.
+type prefixMergeIterator struct {
+	invIt       rocks.EntryIterator
+	committedIt rocks.EntryIterator
+
+	err      error
+	bothDone bool
+}
+
+func newPrefixMergeIterator(iterator rocks.EntryIterator, it rocks.EntryListingIterator, prefixes []string) rocks.EntryIterator {
+	return &prefixMergeIterator{
+		invIt:       iterator,
+		committedIt: newIgnorePrefixIterator(it, prefixes),
+		err:         nil,
+		bothDone:    false,
+	}
+}
+
+func (pmi *prefixMergeIterator) Next() bool {
+	if pmi.err != nil {
+		return false
+	}
+
+	switch {
+	case pmi.invIt.Value() == nil && pmi.committedIt.Value() == nil:
+		pmi.bothDone = true
+		return false
+
+	case pmi.invIt.Value() == nil:
+		return pmi.advanceIt(pmi.committedIt)
+
+	case pmi.committedIt.Value() == nil:
+		return pmi.advanceIt(pmi.invIt)
+
+		// from now on - both iterator has values,
+		// so we'll always return true
+	case strings.Compare(pmi.committedIt.Value().Path.String(), pmi.invIt.Value().Path.String()) < 0:
+		_ = pmi.advanceIt(pmi.committedIt)
+		return true
+
+	default:
+		_ = pmi.advanceIt(pmi.invIt)
+		return true
+	}
+}
+
+func (pmi *prefixMergeIterator) advanceIt(it rocks.EntryIterator) bool {
+	hasNext := it.Next()
+	pmi.err = it.Err()
+	return hasNext
+}
+
+var ErrNotSeekable = errors.New("iterator isn't seekable")
+
+func (pmi *prefixMergeIterator) SeekGE(_ rocks.Path) {
+	pmi.err = ErrNotSeekable
+}
+
+func (pmi *prefixMergeIterator) Value() *rocks.EntryRecord {
+	if pmi.err != nil || pmi.bothDone {
+		return nil
+	}
+
+	committedVal := pmi.committedIt.Value()
+	invVal := pmi.invIt.Value()
+
+	if committedVal == nil {
+		return invVal
+	}
+	if invVal == nil {
+		return committedVal
+	}
+
+	if strings.Compare(committedVal.Path.String(), invVal.Path.String()) < 0 {
+		return committedVal
+	}
+
+	return invVal
+}
+
+func (pmi *prefixMergeIterator) Err() error {
+	return pmi.err
+}
+
+func (pmi *prefixMergeIterator) Close() {
+	pmi.invIt.Close()
+	pmi.err = pmi.invIt.Err()
+}
+
+type ignorePrefixIterator struct {
```

Comment:
I think this iterator needs a test.  Doesn't have to be in this PR, but should be before the upcoming release.

Reply:
I will add as part of this PR.